controllers/coordination/name_registry.go

```diff
@@ -0,0 +1,123 @@
+	coordinationv1 "k8s.io/api/coordination/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	hashedNamePrefix = "n-"
+
+	EntityTypeAnnotation = "coordination.cloudfoundry.org/entity-type"
+	NamespaceAnnotation  = "coordination.cloudfoundry.org/namespace"
+	NameAnnotation       = "coordination.cloudfoundry.org/name"
+)
+
+var (
+	unlockedIdentity = "none"
+	lockedIdentity   = "locked"
+)
+
+//go:generate go run github.com/maxbrunsfeld/counterfeiter/v6 -generate
+//counterfeiter:generate -o fake -fake-name Client sigs.k8s.io/controller-runtime/pkg/client.Client
+
+type NameRegistry struct {
+	client     client.Client
+	entityType string
+}
+
+func NewNameRegistry(client client.Client, entityType string) NameRegistry {
+	return NameRegistry{
+		client:     client,
+		entityType: entityType,
+	}
+}
+
+func (r NameRegistry) RegisterName(ctx context.Context, namespace, name string) error {
+	lease := &coordinationv1.Lease{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      hashName(r.entityType, name),
+			Namespace: namespace,
+			Annotations: map[string]string{
+				EntityTypeAnnotation: r.entityType,
+				NamespaceAnnotation:  namespace,
+				NameAnnotation:       name,
+			},
+		},
+		Spec: coordinationv1.LeaseSpec{
+			HolderIdentity: &unlockedIdentity,
+		},
+	}
+
+	if err := r.client.Create(ctx, lease); err != nil {
+		return fmt.Errorf("creating a lease failed: %w", err)
+	}
+
+	return nil
+}
+
+func (r NameRegistry) DeregisterName(ctx context.Context, namespace, name string) error {
+	lease := &coordinationv1.Lease{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      hashName(r.entityType, name),
+			Namespace: namespace,
+		},
+	}
+	if err := r.client.Delete(ctx, lease); err != nil {
+		return fmt.Errorf("deleting a lease failed: %w", err)
+	}
+
+	return nil
+}
+
+func (r NameRegistry) TryLockName(ctx context.Context, namespace, name string) error {
+	lease := &coordinationv1.Lease{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      hashName(r.entityType, name),
+			Namespace: namespace,
+		},
+	}
+	jsonPatch := fmt.Sprintf(`[
+    {"op":"test", "path":"/spec/holderIdentity", "value": "%s"},
+    {"op":"replace", "path":"/spec/holderIdentity", "value": "%s"}
+    ]`, unlockedIdentity, lockedIdentity)
+
+	if err := r.client.Patch(ctx, lease, client.RawPatch(types.JSONPatchType, []byte(jsonPatch))); err != nil {
+		return fmt.Errorf("failed to acquire lock on lease: %w", err)
+	}
+
+	return nil
+}
+
+func (r NameRegistry) UnlockName(ctx context.Context, namespace, name string) error {
+	lease := &coordinationv1.Lease{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      hashName(r.entityType, name),
+			Namespace: namespace,
+		},
+	}
+
+	copy := lease.DeepCopy()
+	copy.Spec.HolderIdentity = &unlockedIdentity
+
+	if err := r.client.Patch(ctx, copy, client.MergeFrom(lease)); err != nil {
```

Comment:
@georgethebeatle and I were wondering whether we should allow unlocking names that are not locked, i.e. use the `test-replace` patch operation similar to the one in `TryLockName`. 
On one hand, not doing that makes the unlock operation idempotent, on the other, if you are unlocking an unlocked object, that might be an unexpected state (like when you find your car opened, you would start wondering whether someone else has the keys :))
What do you think?

Reply:
This seems reasonable. We've made the change :)